### ISSUE TITLE
Unable to upgrade mongodb community resource

### ISSUE THREAD

**Issue by tnormani** [mongodb/mongodb-kubernetes-operator#1222]
Hi,

we need to upgrade MongoDB community from version 4.2.22 to 6.0.3, using the operator in business continuity (rolling upgrade).
In details:

1.  we upgraded the operator to the latest version 0.7.8. The upgrade terminated successfully
2. starting from this resource

```yaml
apiVersion: mongodbcommunity.mongodb.com/v1
kind: MongoDBCommunity
metadata:
  name: test-mongodb
spec:
  type: ReplicaSet
  version: "4.2.22"
  members: 3
  security:
    authentication:
      modes: [ "SCRAM", "SCRAM-SHA-256", "SCRAM-SHA-1" ]
  users:
    - name: dbroot
      db: admin
      passwordSecretRef:
        name: test-dbroot-password
      roles:
        - name: dbAdmin
          db: admin
        - name: clusterAdmin
          db: admin
        - name: readWriteAnyDatabase
          db: admin
        - name: userAdminAnyDatabase
          db: admin
        - name: dbAdminAnyDatabase
          db: admin
      scramCredentialsSecretName: test-mongodb-admin-dbroot
  statefulSet:
    spec:
      selector:
        matchLabels:
          app: test-mongodb-svc
      # Specifies a size for the data amd logs volumes
      volumeClaimTemplates:
        - metadata:
            name: data-volume
          spec:
            accessModes: [ "ReadWriteOnce" ]
            resources:
              requests:
                storage: 10Gi
        - metadata:
            name: logs-volume
          spec:
            accessModes: [ "ReadWriteOnce" ]
            resources:
              requests:
                storage: 2Gi
      template:
        metadata:
          labels:
            app: test-mongodb-svc
        spec:
          affinity:
            podAntiAffinity:
              requiredDuringSchedulingIgnoredDuringExecution:
                - labelSelector:
                    matchExpressions:
                      - key: app
                        operator: In
                        values:
                          - test-mongodb-svc
                  topologyKey: "kubernetes.io/hostname"
          containers:
            - name: mongod
              resources:
                limits:
                  memory: 1Gi
                requests:
                  cpu: 1200m
            - name: mongodb-agent
              resources:
                limits:
                  memory: 256Mi
                requests:
                  cpu: 50m
              readinessProbe:
                initialDelaySeconds: 10
                periodSeconds: 10
                timeoutSeconds: 10
                successThreshold: 1
                failureThreshold: 10
```
which is deployed **successfully** 

3.  we moved to 
```yaml
apiVersion: mongodbcommunity.mongodb.com/v1
kind: MongoDBCommunity
metadata:
  name: test-mongodb
spec:
  type: ReplicaSet
  version: "4.4.18"
  featureCompatibilityVersion: "4.2"
  members: 3
  security:
    authentication:
      modes: [ "SCRAM", "SCRAM-SHA-256", "SCRAM-SHA-1" ]
  users:
    - name: dbroot
      db: admin
      passwordSecretRef:
        name: test-dbroot-password
      roles:
        - name: dbAdmin
          db: admin
        - name: clusterAdmin
          db: admin
        - name: readWriteAnyDatabase
          db: admin
        - name: userAdminAnyDatabase
          db: admin
        - name: dbAdminAnyDatabase
          db: admin
      scramCredentialsSecretName: test-mongodb-admin-dbroot
  statefulSet:
    spec:
      selector:
        matchLabels:
          app: test-mongodb-svc
      # Specifies a size for the data amd logs volumes
      volumeClaimTemplates:
        - metadata:
            name: data-volume
          spec:
            accessModes: [ "ReadWriteOnce" ]
            resources:
              requests:
                storage: 10Gi
        - metadata:
            name: logs-volume
          spec:
            accessModes: [ "ReadWriteOnce" ]
            resources:
              requests:
                storage: 2Gi
      template:
        metadata:
          labels:
            app: test-mongodb-svc
        spec:
          affinity:
            podAntiAffinity:
              requiredDuringSchedulingIgnoredDuringExecution:
                - labelSelector:
                    matchExpressions:
                      - key: app
                        operator: In
                        values:
                          - test-mongodb-svc
                  topologyKey: "kubernetes.io/hostname"
          containers:
            - name: mongod
              resources:
                limits:
                  memory: 1Gi
                requests:
                  cpu: 1200m
            - name: mongodb-agent
              resources:
                limits:
                  memory: 256Mi
                requests:
                  cpu: 50m
              readinessProbe:
                initialDelaySeconds: 10
                periodSeconds: 10
                timeoutSeconds: 10
                successThreshold: 1
                failureThreshold: 10
```
This upgrade isn't coming to an end. After several minutes, the situation is the following
```shell
$ kubectl get pods -n mongodb -w
NAME                                           READY   STATUS    RESTARTS   AGE
mongodb-kubernetes-operator-55d488798b-54m87   1/1     Running   0          33m
test-mongodb-0                                2/2     Running   0          23m
test-mongodb-1                                1/2     Running   0          32m
```
test-mongodb-0 is upgraded successfully, test-mongodb-1 is still 4.2 version and agent-probe is not ready and test-mongodb-2 is missing totally. The situation doesn't seem to change after hours. Manually deleting test-mongodb-1 seems to solve the stale situation, but it breaks business continuity process.

**Operator Information**
 - Operator Version: 0.7.8
 - MongoDB Image used: 
   - quay.io/mongodb/mongodb-kubernetes-readinessprobe:1.0.12
   - quay.io/mongodb/mongodb-agent:12.0.15.7646-1
   - docker.io/mongo:4.4.18

**Kubernetes Cluster Information**
 - Distribution: Anthos 1.13, Kind 0.17.0, Minikube 1.28.0
 - Image Registry location (quay, or an internal registry): quay

**Additional context**
No matter K8s cluster flavour or the version of the K8s itself, the behaviour described is the same. We also tried various MongoDB upgrades and the behaviour is the same too.

Tell us what we are doing wrong, because it's very unlikely that such an important feature is broken.
Thank you in advance

**Comment by durera** [comment]
> Tell us what we are doing wrong, because it's very unlikely that such an important feature is broken.

Unfortunately I can't tell you what you're doing wrong ... only that whatever you are doing wrong is exactly what we are doing wrong too :)

We have been able to upgrade the operator from 0.7 to 0.7.8, but any attempt to change the version in the MongoDBCommunity spec leads us to the same position you've described where the cluster can not come back up.

It's not just limited to minor version bumps .. when testing 4.2.x  to 4.2.23 we hit the same problems ... but changing the version while staying on the 0.7.0 operator works fine, so it may well be a bug with the 0.7.8 operator .. or some extra step that's needed which both of us have missed.

## Starting point
An old 4.2.6 install on 0.7.0 operator:
```
oc -n mongoce get pods,sts,mongodbcommunity
NAME                                               READY   STATUS    RESTARTS   AGE
pod/mas-mongo-ce-0                                 2/2     Running   0          9m4s
pod/mas-mongo-ce-1                                 2/2     Running   0          8m4s
pod/mas-mongo-ce-2                                 2/2     Running   0          7m4s
pod/mongodb-kubernetes-operator-59f757c655-8fm9f   1/1     Running   0          9m20s

NAME                            READY   AGE
statefulset.apps/mas-mongo-ce   3/3     9m4s

NAME                                                         PHASE     VERSION
mongodbcommunity.mongodbcommunity.mongodb.com/mas-mongo-ce   Running
```

## Test 1
Update to 4.2.23 after updating to 0.7.8 operator: fail
```
NAME                                               READY   STATUS    RESTARTS   AGE
pod/mas-mongo-ce-0                                 1/2     Running   0          47m
pod/mas-mongo-ce-1                                 1/2     Running   0          46m
pod/mas-mongo-ce-2                                 1/2     Running   0          33m
pod/mongodb-kubernetes-operator-78ffc979b9-9l6cj   1/1     Running   0          33m

NAME                                READY   AGE
statefulset.apps/mas-mongo-ce       0/3     47m
statefulset.apps/mas-mongo-ce-arb   0/0     33m

NAME                                                         PHASE     VERSION
mongodbcommunity.mongodbcommunity.mongodb.com/mas-mongo-ce   Pending
```

## Test 2
Update to 4.2.23, staying on 0.7.0 operator: success
```
NAME                                               READY   STATUS    RESTARTS   AGE
pod/mas-mongo-ce-0                                 2/2     Running   0          7m27s
pod/mas-mongo-ce-1                                 2/2     Running   0          10m
pod/mas-mongo-ce-2                                 2/2     Running   0          9m1s
pod/mongodb-kubernetes-operator-59f757c655-ctnhm   1/1     Running   0          18m

NAME                            READY   AGE
statefulset.apps/mas-mongo-ce   3/3     17m

NAME                                                         PHASE     VERSION
mongodbcommunity.mongodbcommunity.mongodb.com/mas-mongo-ce   Running
```

**Comment by terenceq** [comment]
Following the steps outlined here:

https://github.com/mongodb/mongodb-kubernetes-operator/blob/master/docs/install-upgrade.md#upgrade-the-operator

seem to work (at least when leveraging MongoDB 4.2.x containers). Deleting the operator deployment and StatefulSet before upgrading the operator works with my tests.

**Comment by tzorzan** [comment]
Here we are talking about upgrading the _replicaset_ version, from MongoDB 4.2.x to 6.0 without downtime. The operator is already at 0.7.8 and doesn't need to be updated.

**Comment by dan-mckean** [comment]
Hi, I'm Dan and I'm the Product Manager for MongoDB's support of Kubernetes.

I'm doing some work right now to try and identify how the Community Operator is being used. The Community Operator is something I inherited when I started at MongoDB, but it doesn't get as much attention from us as we'd like and we're trying to understand how it's used in order to establish its future. It will help us prioritize future issues and PRs raised by the community 🙂

Here's a super short survey (it's much easier for us to review all the feedback that way!): [https://docs.google.com/forms/d/e/1FAIpQLSfwrwyxBSlUyJ6AmC-eYlgW_3JEdfA48SB2i5--_WpiynMW2w/viewform?usp=sf_link](https://docs.google.com/forms/d/e/1FAIpQLSfwrwyxBSlUyJ6AmC-eYlgW_3JEdfA48SB2i5--_WpiynMW2w/viewform?usp=sf_link)

 If you'd rather email me instead: [dan.mckean@mongodb.com](mailto:dan.mckean@mongodb.com?subject=MongoDB%20Community%20Operator%20feedback)

Thank you in advance!
Dan
